### PR TITLE
Converting core tests to use testable fake implementations

### PR DESCRIPTION
**Pull request by hmemcpy**
Connects to https://github.com/Particular/NServiceBus.Testing/issues/29

### CODE REVIEW

File: src/NServiceBus.Testing.Fakes/NServiceBus.Testing.Fakes.csproj [hunk 1]
```diff
@@ -79,6 +85,9 @@
     <None Include="..\..\packaging\nuget\nservicebus.testing.fakes.nuspec">
       <Link>nservicebus.testing.fakes.nuspec</Link>
     </None>
+    <None Include="..\Test.snk">
+      <Link>Test.snk</Link>
```
Comment (andreasohlund) [inline src/NServiceBus.Testing.Fakes/NServiceBus.Testing.Fakes.csproj]: Do we need to strongname? (what internals do the core test need access to?)

Comment (hmemcpy) [inline src/NServiceBus.Testing.Fakes/NServiceBus.Testing.Fakes.csproj]: We do, because it's a project reference. Core.Tests are strong named, and it won't compile otherwise.

Comment (andreasohlund) [inline src/NServiceBus.Testing.Fakes/NServiceBus.Testing.Fakes.csproj]: :+1:

File: src/NServiceBus.Core.Tests/Scheduler/ScheduledTaskMessageHandlerTests.cs [hunk 2]
```diff
@@ -36,7 +37,8 @@ public void When_a_scheduledTask_message_is_handled_the_task_should_be_defer()
                 TaskId = taskId
             }, handlingContext);
 
-            Assert.That(((ScheduledTask)handlingContext.DeferedMessage).TaskId, Is.EqualTo(taskId));
+            var deferredMessage = handlingContext.SentMessages.First(message => message.Options.GetDeliveryDelay().HasValue).Message;
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Scheduler/ScheduledTaskMessageHandlerTests.cs]: I think you can use `var deferredMessage = handlingContext.SentMessages.First(message => message.Options.GetDeliveryDelay().HasValue).Message<ScheduledTask>();` to avoid casting on the next line. Or the `Containing` extension method.

**Comment by WojcikMike** [comment]
## Status
I went through a lot of test in core (from the top to end of Routing directory) and did some changes to it. During this changes I identified few good, bad and ugly changes that using Fakes implies.

### Good cases

See changes in `MulticastPublishRouterBehaviorTests.cs` file. Becasue of fakes we were able to remove rouge implementation of FakeContext. So that is clear simplification. 

Second example is `MutateOutgoingMessageBehaviorTests.cs` in which below code:

```
            var message = new FakeMessage();

            var outgoingLogicalMessage = new OutgoingLogicalMessage(message.GetType(),message);

            var logicalContext = new OutgoingLogicalMessageContext("messageId", new Dictionary<string, string>(), outgoingLogicalMessage, new List<RoutingStrategy>(), null);

            var physicalContext = new OutgoingPhysicalMessageContext(new byte[0], new List<RoutingStrategy>(), logicalContext);

            physicalContext.Set(outgoingLogicalMessage);

            var builder = new FuncBuilder();

            builder.Register<IMutateOutgoingTransportMessages>(() => new MutateOutgoingTransportMessagesReturnsNull());

            physicalContext.Set<IBuilder>(builder);
``` 

was replaced with:
```
			var context = new TestableOutgoingLogicalMessageContext();
            context.Extensions.Set(new IncomingMessage("messageId", new Dictionary<string, string>(), Stream.Null ));
            context.Extensions.Set(new LogicalMessage(null, null));
            context.Builder.Register<IMutateOutgoingMessages>(() => new MutateOutgoingMessagesReturnsNull());
```

### Ugly Cases

I would consider them as having almost no impact on test. Example would be `RouteDeferredMessageToTimeoutManagerBehaviorTests.cs` where code:

```
	var context = new RoutingContext(message, new UnicastRoutingStrategy("target"), null);
    context.AddDeliveryConstraint(new DelayDeliveryWith(delay));
```
Is replaced with:
```
	var context = new TestableRoutingContext
            {
                RoutingStrategies = new List<RoutingStrategy>
                {
                    new UnicastRoutingStrategy("target")
                }
            };
    context.Extensions.AddDeliveryConstraint(new DelayDeliveryWith(delay));
```
It is mostly the reason of needing to modify some properties like adding routing strategy in this case.


### Bad Cases
This are the cases that at the moment we can't replace with Fakes. `MoveFaultsToErrorQueueTests`, `FirstLevelRetriesTests` and `SecondLevelRetriesTests`. The reason for that is that they use `FakeBehaviorContext` which uses `FakeEventAggregator` which we don't have in fakes.

File: src/NServiceBus.Core.Tests/DelayedDelivery/RouteDeferredMessageToTimeoutManagerBehaviorTests.cs [hunk 3]
```diff
@@ -18,12 +18,17 @@ class RouteDeferredMessageToTimeoutManagerBehaviorTests
...
             var delay = TimeSpan.FromDays(1);
-            var message = new OutgoingMessage("id", new Dictionary<string, string>(), new byte[0]);
 
             var headers = new Dictionary<string, string>();
             string destination = null;
-            var context = new RoutingContext(message, new UnicastRoutingStrategy("target"), null);
-            context.AddDeliveryConstraint(new DelayDeliveryWith(delay));
+            var context = new TestableRoutingContext
+            {
+                RoutingStrategies = new List<RoutingStrategy>
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/DelayedDelivery/RouteDeferredMessageToTimeoutManagerBehaviorTests.cs]: Isn't this a sign that https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Pipeline/Outgoing/IRoutingContext.cs#L20 shouldn't be a readonly collection?

If it's readonly why is there a setter?

File: src/NServiceBus.Core.Tests/NServiceBus.Core.Tests.csproj [hunk 4]
```diff
@@ -372,5 +375,6 @@
   <ItemGroup>
     <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
   </ItemGroup>
+  <ItemGroup />
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/NServiceBus.Core.Tests.csproj]: Remove?

File: src/NServiceBus.Core.Tests/Reliability/Outbox/TransportReceiveToPhysicalMessageProcessingConnectorTests.cs [hunk 5]
```diff
@@ -108,9 +111,12 @@ public class TransportReceiveToPhysicalMessageProcessingConnectorTests
             Assert.Null(fakeOutbox.StoredMessage);
         }
 
-        static ITransportReceiveContext CreateContext(FakeBatchPipeline pipeline)
+        static ITransportReceiveContext CreateContext(FakeBatchPipeline pipeline, string messageId)
         {
-            var context = new TransportReceiveContext(new IncomingMessage("id", new Dictionary<string, string>(), new MemoryStream()), null, new CancellationTokenSource(), new RootContext(null, new FakePipelineCache(pipeline), null));
+            var context = new TestableTransportReceiveContext() {Message = new IncomingMessage(messageId, new Dictionary<string, string>(), Stream.Null) };
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Reliability/Outbox/TransportReceiveToPhysicalMessageProcessingConnectorTests.cs]: Formatting

File: src/NServiceBus.Core.Tests/Routing/ApplyReplyToAddressBehaviorTests.cs [hunk 6]
```diff
@@ -25,17 +25,10 @@ public class ApplyReplyToAddressBehaviorTests
...
-            var context = new OutgoingLogicalMessageContext(
-                Guid.NewGuid().ToString(),
-                new Dictionary<string, string>(),
-                new OutgoingLogicalMessage(typeof(MyMessage), new MyMessage()),
-                new RoutingStrategy[]
-                {
-                },
-                new RootContext(null, null, null));
-            context.Merge(options.Context);
+            var context = new TestableOutgoingLogicalMessageContext {Extensions = options.Context};
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Routing/ApplyReplyToAddressBehaviorTests.cs]: Formatting

File: src/NServiceBus.Core.Tests/Routing/DetermineRouteForReplyBehaviorTests.cs [hunk 7]
```diff
@@ -18,20 +18,16 @@ public class DetermineRouteForReplyBehaviorTests
...
-                        },
-                        new MemoryStream()), null, new CancellationTokenSource(),
-                    new RootContext(null, null, null)));
+            var context = CreateContext(new OutgoingLogicalMessage(typeof(MyReply), new MyReply()));
+
+            context.Extensions.Set(new IncomingMessage(
+                "id",
+                new Dictionary<string, string>
+                {
+                    {Headers.ReplyToAddress, "ReplyAddressOfIncomingMessage"}
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Routing/DetermineRouteForReplyBehaviorTests.cs]: Formatting

File: src/NServiceBus.Core.Tests/Routing/MessageDrivenUnsubscribeTerminatorTests.cs [hunk 8]
```diff
@@ -44,7 +45,10 @@ public void SetUp()
             state.RetryDelay = TimeSpan.Zero;
             dispatcher.FailDispatch(10);
 
-            await terminator.Invoke(new UnsubscribeContext(new FakeContext(), typeof(object), options), c => TaskEx.CompletedTask);
+            var context = new TestableUnsubscribeContext();
+            context.Extensions = options.Context;
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenUnsubscribeTerminatorTests.cs]: Use initializer syntax?

File: src/NServiceBus.Core.Tests/Routing/MessageDrivenUnsubscribeTerminatorTests.cs [hunk 9]
```diff
@@ -59,7 +63,10 @@ public void Should_Throw_when_max_retries_reached()
             state.RetryDelay = TimeSpan.Zero;
             dispatcher.FailDispatch(11);
 
-            Assert.That(async () => await terminator.Invoke(new UnsubscribeContext(new FakeContext(), typeof(object), options), c => TaskEx.CompletedTask), Throws.InstanceOf<QueueNotFoundException>());
+            var context = new TestableUnsubscribeContext();
+            context.Extensions = options.Context;
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenUnsubscribeTerminatorTests.cs]: Initializer syntax?

File: src/NServiceBus.Core.Tests/Routing/RoutingToDispatchConnectorTests.cs [hunk 10]
```diff
@@ -15,13 +16,10 @@ public class RoutingToDispatchConnectorTests
         public async Task Should_preserve_headers_generated_by_custom_routing_strategy()
         {
             var behavior = new RoutingToDispatchConnector();
-            var message = new OutgoingMessage("ID", new Dictionary<string, string>(), new byte[0]);
             Dictionary<string, string> headers = null;
-
-            await behavior.Invoke(new RoutingContext(message,
-                new CustomRoutingStrategy(),CreateContext(new SendOptions(), false)), c =>
+            await behavior.Invoke(new TestableRoutingContext() {RoutingStrategies = new List<RoutingStrategy> { new CustomRoutingStrategy() } }, context =>
```
Comment (andreasohlund) [inline src/NServiceBus.Core.Tests/Routing/RoutingToDispatchConnectorTests.cs]: Formatting (space before "RoutingStrategies = ")

File: src/NServiceBus.Core.Tests/DelayedDelivery/RouteDeferredMessageToTimeoutManagerBehaviorTests.cs [hunk 11]
```diff
@@ -71,12 +84,16 @@ public void Cannot_be_combined_with_time_to_be_received()
...
             var behavior = new RouteDeferredMessageToTimeoutManagerBehavior("tm");
             var delay = TimeSpan.FromDays(1);
-
-            var message = new OutgoingMessage("id", new Dictionary<string, string>(), new byte[0]);
-
+            
             var headers = new Dictionary<string, string>();
-            var context = new RoutingContext(message, new UnicastRoutingStrategy("target"), null);
-            context.AddDeliveryConstraint(new DelayDeliveryWith(delay));
+            var context = new TestableRoutingContext
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/DelayedDelivery/RouteDeferredMessageToTimeoutManagerBehaviorTests.cs]: in situations like this it would probably make sense to introduce a local "helper" method to reduce the noise
e.g. `CreateRoutingContext(...)`

File: src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/DispatchTimeoutBehaviorTest.cs [hunk 12]
```diff
@@ -119,8 +120,7 @@ static ISatelliteProcessingContext CreateContext(string timeoutId)
                 {"Timeout.Id", timeoutId}
             };
 
-            return new SatelliteProcessingContext(
-                new IncomingMessage(messageId, headers, new MemoryStream()), null);
+            return new TestableSatelliteProcessingContext {Message = new IncomingMessage(messageId, headers, new MemoryStream()) };
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/DelayedDelivery/TimeoutManager/DispatchTimeoutBehaviorTest.cs]: (you can use `Stream.Null` instead of `new MemoryStream()`)

File: src/NServiceBus.Core.Tests/NServiceBus.Core.Tests.csproj (hunk 4 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/NServiceBus.Core.Tests.csproj]: can that be removed?

File: src/NServiceBus.Core.Tests/Pipeline/Incoming/InvokeHandlerTerminatorTest.cs [hunk 13]
```diff
@@ -162,15 +163,10 @@ static MessageHandler CreateMessageHandlerThatReturnsNull(Action<object, object,
...
         {
-            var behaviorContext = new InvokeHandlerContext(
-                messageHandler,
-                "messageId",
-                "replyToAddress",
-                new Dictionary<string, string>(),
-                null,
-                null,
-                null,
-                new RootContext(null, null, null));
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Pipeline/Incoming/InvokeHandlerTerminatorTest.cs]: :tada:

File: src/NServiceBus.Core.Tests/Routing/DetermineRouteForReplyBehaviorTests.cs [hunk 14]
```diff
@@ -18,20 +18,16 @@ public class DetermineRouteForReplyBehaviorTests
...
+                },
+                new MemoryStream()));
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Routing/DetermineRouteForReplyBehaviorTests.cs]: Stream.Null

File: src/NServiceBus.Core.Tests/Routing/DetermineRouteForReplyBehaviorTests.cs [hunk 15]
```diff
@@ -47,17 +43,13 @@ public class DetermineRouteForReplyBehaviorTests
...
-                        "id",
-                        new Dictionary<string, string>(),
-                        new MemoryStream()), null, new CancellationTokenSource(),
-                    new RootContext(null, null, null)));
+            var context = CreateContext(new OutgoingLogicalMessage(typeof(MyReply), new MyReply()));
+
+            context.Extensions.Set(new IncomingMessage(
+                "id",
+                new Dictionary<string, string>(),
+                new MemoryStream()));
```
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Routing/DetermineRouteForReplyBehaviorTests.cs]: Stream.Null

File: src/NServiceBus.Core.Tests/Routing/MessageDrivenUnsubscribeTerminatorTests.cs (hunk 9 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenUnsubscribeTerminatorTests.cs]: this seems to be not that uncommon. Would it be interesting to provide some helper to merge SendOptions into the context? e.g. `context.Merge(sendOptions);`

File: src/NServiceBus.Testing.Fakes/TestableOutgoingContext.cs [hunk 16]
```diff
@@ -17,7 +17,9 @@ public partial class TestableOutgoingContext : TestablePipelineContext, IOutgoin
         /// </summary>
         public FakeBuilder Builder { get; set; } = new FakeBuilder();
 
-        IBuilder IBehaviorContext.Builder { get; }
+        IBuilder IBehaviorContext.Builder {
+            get { return this.Builder; }
```
Comment (timbussmann) [inline src/NServiceBus.Testing.Fakes/TestableOutgoingContext.cs]: this is already fixed on latest develop. Best to revert and rebase (it should call GetBuilder() to enable users to provide their custom IBuilder).

File: src/NServiceBus.Core.Tests/DelayedDelivery/RouteDeferredMessageToTimeoutManagerBehaviorTests.cs (hunk 3 above)
Comment (timbussmann) [inline src/NServiceBus.Core.Tests/DelayedDelivery/RouteDeferredMessageToTimeoutManagerBehaviorTests.cs]: `RouteDeferredMessageToTimeoutManagerBehavior` seems to be the only place using that setter to replace the whole collection.

But that seems to be a bit out of scope?

Comment (WojcikMike) [inline src/NServiceBus.Core.Tests/DelayedDelivery/RouteDeferredMessageToTimeoutManagerBehaviorTests.cs]: We use setter as TestableRoutingContext create RoutingStrategies automatically, if you want to provide your own you need to use setter. I think it is acceptable within tests and only in TestableRoutincContext there is a setter for that.

File: src/NServiceBus.Core.Tests/Routing/MessageDrivenUnsubscribeTerminatorTests.cs (hunk 9 above)
Comment (WojcikMike) [inline src/NServiceBus.Core.Tests/Routing/MessageDrivenUnsubscribeTerminatorTests.cs]: you can write `context.Extensions.Merge` for that. But as I don't have anything in Extensions I wanted to keep test simple and limit calling methods

**Comment by timbussmann** [comment]
addressed the "bad" cases and removed the `FakeBehaviorContext`

**Comment by WojcikMike** [comment]
Comments addressed, commits squashed and WIP removed. please merge @Particular/nservicebus-maintainers

**Comment by andreasohlund** [comment]
2 unit tests failing

**Comment by WojcikMike** [comment]
fixed
